### PR TITLE
Make message editable

### LINKED ISSUE #4: Make message editable
By default, this block shows the message `Spend only $xx.xx more to get free US shipping`. To enable merchants to change the info message according to their needs, that field needs to be editable. However, the remaining amount is dynamic and will be calculated based on the free shipping value and the added products. The merchants must also be able to edit both messages, when the free shipping value hasn't been reached yet and when it has been reached.

We need to find a way to empower merchants to edit the message while ensuring that the variable with the remaining amount remains working.

## Possible solutions

- Add a placeholder to the message, e.g. `Spend only {amount} more to get free US shipping`.
- Make only the parts `Spend` and `more to get free US shipping` editable.
- Make the text editable using `<TextControl>` within `<InspectorControls>`.

## Current translations

- `Spend only {amount} more to get free US shipping!`
- `You have qualified for free shipping. Great job!`

### PR DESCRIPTION
**Pull request by nielslange**
Fixes #4 

## Note

In i1 of this block, the merchant must be able to edit the messages for insufficient and sufficient cart totals. This PR adds this functionality.

<table>
<tr>
<td>Before:
<br><br>

#4-before
</td>
<td>After (separated textareas):
<br><br>

#4-after-seperated
</td>
<td>After (combined textarea):
<br><br>

#4-after-combined
</td>
</tr>
</table>

To enable merchants to adjust the message for insufficient cart totals, e.g. `Spend only $35.00 more to get free US shipping!`, according to their needs, I initially created two textareas. One textarea for the message before the remaining value and another one for the message after the remaining value. To keep the editing of the messages simple, I decided against that solution and created just one textarea to adjust the message for insufficient cart totals. This message includes the placeholder `%s`. In the help text, I highlighted the importance of this placeholder.

I created the two tags [Insufficient-totals-with-separate-textareas](https://github.com/woocommerce/woocommerce-free-shipping-progress-bar-block/releases/tag/Insufficient-totals-with-seperate-textareas) and [Insufficient-totals-with-combined-textarea](https://github.com/woocommerce/woocommerce-free-shipping-progress-bar-block/releases/tag/Insufficient-totals-with-one-textarea) so that both versions can be checked during the code review.

## Testing instructions

1. Check out this PR.
2. Create a test page.
3. Add the Free Shipping Progress Bar block to it and save the page.
4. Mark the block and open the sidebar.
5. Change the `Insufficient totals message` and `Sufficient totals message`.
6. Change the `Free shipping from` value to see both messages.
7. Ensure that the changes are visible both in the editor and the frontend.

### CODE REVIEW

File: src/edit.js [hunk 1]
```diff
@@ -26,20 +27,67 @@ import Block from './block';
...
+		insufficientTotals,
+		sufficientTotals,
+		freeShippingFrom,
+	} = attributes;
 
 	return (
 		<div { ...useBlockProps() }>
 			<InspectorControls>
-				<PanelBody title="Settings" initialOpen={ true }>
+				<PanelBody
```
Comment (senadir) [inline src/edit.js]: Thank you for working on this Niels, unfortunately, this is not how we want to implement text editing.

All text editing in Gutenberg should be inline. It doesn't make sense to have it in the sidebar. To do this, you will need to break down your block into two components, a progress bar, and the text part, one that is editable in the editor, the second, static, in frontend. We did this in the Product Search component for example.

You also need to be able to edit two views, so you need a way to toggle between them, to do that, you can check the Cart i2 block for how we implement it, it's a bit complicated there, but all you need to do is:
- A Toolbar and ToolbarDropdownMenu that toggles between values that you save in state.
- You use what's in state to show the correct text.

File: src/edit.js [hunk 2]
```diff
@@ -26,20 +27,67 @@ import Block from './block';
...
+					title={ __( 'Settings', 'free-shipping-progress-bar' ) }
+					initialOpen={ true }
+				>
+					<TextareaControl
+						label={ __(
+							'Insufficient totals message',
+							'free-shipping-progress-bar'
+						) }
+						help={ __(
+							'Provide the message for insufficient totals. %s is a placeholder for the calculated remaining value and should not be removed.',
```
Comment (senadir) [inline src/edit.js]: Consider using something more obvious than %s. Consider showing a notice if %s isn't present in the text. We do this in the Terms and Conditions block in Checkout, in which we show an error if the text doesn't contain a string, and offer a reset.

File: src/progress-bar-icon.js [hunk 3]
```diff
@@ -48,16 +48,16 @@ export default function ProgressBarIcon() {
 							xlinkHref="#a3HIePHoOc"
 							opacity="1"
 							fill="#96588a"
-							fill-opacity="0"
+							fillOpacity="0"
```
Comment (senadir) [inline src/progress-bar-icon.js]: This is again present in this PR, why?
I think you should move this code to another PR, have it reviewed and merged, then rebase this and the colors PR against trunk.

Also this PR still has `wp-element` as a static value in php.

**Review by senadir — CHANGES_REQUESTED** [review]
Hey Niels. Thank you for working on this, I left some changes because this is not how we offer text editing.

File: src/progress-bar-icon.js (hunk 3 above)
Comment (nielslange) [reply src/progress-bar-icon.js]: I removed these changes form this PR and from #14.

File: src/edit.js (hunk 1 above)
Comment (nielslange) [reply src/edit.js]: Thanks for your review, @senadir. I'll adjust this PR based on your feedback.

File: src/edit.js (hunk 2 above)
Comment (nielslange) [reply src/edit.js]: Thanks for pointing out the example in the terms and conditions block within the checkout block.

Comment (nielslange) [reply src/edit.js]: @senadir Given that the string can occur in various languages, I'm only showing a warning, but I haven't implemented the action to copy the original string `Spend only {REMAINING_AMOUNT} more to get free US shipping!` back into the `<PlainText>` field. I'm keen to hear your thoughts on this.

File: src/edit.js (hunk 1 above)
Comment (nielslange) [reply src/edit.js]: As requested, I removed the `<TextareaControl>` components from the sidebar and implemented the `<PlainText>` and the `<ViewSwitcher>` components, to enable merchants to edit both the insufficient and sufficient cart total labels inline.

**Comment by nielslange** [comment]
@senadir I've updated this PR according to your feedback and double-checked that the functionality is working as expected, both with Gutenberg activated and deactivated. I'm looking forward to receiving your feedback on this update.

File: block.json [hunk 4]
```diff
@@ -10,12 +10,20 @@
 		"html": false
 	},
 	"attributes": {
+		"labelInsufficientTotals": {
+			"type": "string",
+			"default": "Spend only {REMAINING_AMOUNT} more to get free US shipping!"
```
Comment (senadir) [inline block.json]: Given that if a merchant is going to rewrite this, making sure they write `{REMAINING_AMOUNT}` is going to be hard right? How about you simplify it to `{amount}` or even a non latin symbol like %$%?

File: block.json [hunk 5]
```diff
@@ -10,12 +10,20 @@
...
+		},
+		"labelSufficientTotals": {
+			"type": "string",
+			"default": "You have qualified for free shipping. Great job!"
+		},
 		"currentTotal": {
-			"type": "number",
+			"type": "string",
```
Comment (senadir) [inline block.json:14]: Why did you switch this to a string?

File: src/components/block-error-boundary/block-error.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	text = __(
+		'There was an error loading the content.',
+		'free-shipping-progress-bar'
+	),
+	errorMessage,
+	errorMessagePrefix = __( 'Error:', 'free-shipping-progress-bar' ),
+	button,
+} ) => {
+	return (
+		<div className="wc-free-shipping-progress-bar-error wc-free-shipping-progress-bar-components-error">
```
Comment (senadir) [inline src/components/block-error-boundary/block-error.js]: Why are all classes duplicated here?

File: src/components/progress-label/index.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	freeShippingFrom: PropTypes.string,
+	/**
+	 * The label for when the cart totals are insufficient for free shipping.
+	 */
+	labelInsufficientTotals: PropTypes.string,
+	/**
+	 * The label for when the cart totals are sufficient for free shipping.
+	 */
+	labelSufficientTotals: PropTypes.string,
+};
```
Comment (senadir) [inline src/components/progress-label/index.js]: We moved away from using prop types. Consider using typescript or JSDoc. I'd wait until you move this to a codebase before deciding what to use (based on that codebase).

File: src/components/view-switcher/index.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
+/**
```
Comment (senadir) [inline src/components/view-switcher/index.js]: What instance of viewSwitcher did you copy?

File: src/edit.js [hunk 9]
```diff
@@ -2,43 +2,142 @@
...
-							} )
-						}
-					/>
-				</PanelBody>
-			</InspectorControls>
-			<Block { ...attributes } />
+					>
+						{ currentView === 'insufficient' && (
+							<>
+								<InspectorControls>
```
Comment (senadir) [inline src/edit.js]: Navigating this code is extremely hard now with this level of indention. Can you abstract `InspectorControls` into its own component above and call it here? You can check our Cart and Checkout blocks for examples.

File: src/edit.js [hunk 10]
```diff
@@ -2,43 +2,142 @@
...
+										isDismissible={ false }
+									>
+										<p>
+											{ __(
+												'Ensure that the label contains the placeholder {REMAINING_AMOUNT}.',
+												'free-shipping-progress-bar'
+											) }
+										</p>
+									</Notice>
+								) }
```
Comment (senadir) [inline src/edit.js]: I see your point about not update the value because of a language usage, but you don't have to use a static string here. You can just reuse the default value in you have `block.json` which would be already translated. We do exactly that in Terms and Conditions block.

**Review by senadir — CHANGES_REQUESTED** [review]
Hey Niels, so far this is looking good and in the correct path, however, there are a lot of things left that I pointed out.

File: block.json (hunk 4 above)
Comment (nielslange) [reply block.json]: I went with `{amount}` to avoid that the placeholder is getting to cryptic.

File: block.json (hunk 5 above)
Comment (nielslange) [reply block.json:14]: I understood that we're getting the value of this field as a string from WooCommerce. That's why I changed it. Do you want me to change it back and rather parse the value we're consuming from WooCommerce later on?

File: src/components/block-error-boundary/block-error.js (hunk 6 above)
Comment (nielslange) [reply src/components/block-error-boundary/block-error.js]: I followed the same class structure as in https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/9483e19e52f4fb69615ecdeb5c51394a44f91d19/assets/js/base/components/block-error-boundary/block-error.js#L20-L51. That said, I just removed the duplicate CSS classes.

File: src/components/progress-label/index.js (hunk 7 above)
Comment (nielslange) [reply src/components/progress-label/index.js]: Thanks for pointing this out. I removed the prop types for now and will use TypeScript or JSDoc in the future based on the codebase then.

File: src/edit.js (hunk 9 above)
Comment (nielslange) [reply src/edit.js]: As requested, I moved `<InspectorControls>` into a separate component.

File: block.json (hunk 5 above)
Comment (senadir) [reply block.json:14]: I will come back to this point later on when you start comparing this value against the cart totals.

File: src/components/view-switcher/index.js (hunk 8 above)
Comment (nielslange) [reply src/components/view-switcher/index.js]: I copied the most recent version of the `<ViewSwitcher>` component from https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/343f70d7af57f751c327349dec864f9e176d0f70/assets/js/editor-components/view-switcher/index.js.

File: src/edit.js (hunk 10 above)
Comment (nielslange) [reply src/edit.js]: I just updated this part as well and pushed the change to this PR.

File: src/components/view-switcher/index.js (hunk 8 above)
Comment (senadir) [reply src/components/view-switcher/index.js]: I'm not sure where we discussed this before, but I was talking about the Cart i2 view switcher, so this is the file I'm talking about:
https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/assets/js/blocks/cart-checkout/cart-i2/use-view-switcher.tsx

File: src/constants.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+ * External dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+
+export const notice = sprintf(
+	__(
+		'Spend only {amount} more to get free US shipping!',
+		'free-shipping-progress-bar'
+	)
+);
```
Comment (senadir) [inline src/constants.js]: Why is this a sprintf if you're not replacing anything?

Comment (senadir) [reply src/constants.js]: Also, notice should be the same one used in block.json as well, so you can move the attributes section from block.json to `registerBlockType` section.

**Review by senadir — APPROVED** [review]
This is looking good, I left a feedback to replace viewSwitcher with the newer version, and a note about notice attribute, but all is fine!

File: src/components/view-switcher/index.js (hunk 8 above)
Comment (nielslange) [reply src/components/view-switcher/index.js]: As requested, I replaced `<ViewSwitcher>`.

File: src/constants.js (hunk 11 above)
Comment (nielslange) [reply src/constants.js]: > Why is this a sprintf if you're not replacing anything?

I removed the obsolete `sprintf()` wrapper.

> Also, notice should be the same one used in block.json as well, so you can move the attributes section from block.json to registerBlockType section.

When moving the attributes section from `block.json` to `registerBlockType()`, I run into the following problems:
- I wasn't able to merge attributes from `block.json` and `registerBlockType()`.
- When moving all attributes from `block.json` to `registerBlockType()`, some attributes could not be detected in the frontend.

Therefore, I moved the attributes back to `block.json`.  Would you mind taking a look at this problem?

File: package.json [hunk 12]
```diff
@@ -22,6 +22,7 @@
 	"dependencies": {
 		"@wordpress/block-editor": "^7.0.2",
 		"@wordpress/blocks": "^11.1.0",
+		"@wordpress/components": "^18.0.0",
```
Comment (senadir) [inline package.json]: Those need to move to dev dependencies

File: src/components/use-view-switcher/index.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+ */
+import { Icon, eye } from '../../icons';
+
+export const useViewSwitcher = ( clientId, views ) => {
+	const initialView = views[ 0 ];
+	const [ currentView, setCurrentView ] = useState( initialView );
+	const { selectBlock } = useDispatch( 'core/block-editor' );
+	const { getBlock } = select( blockEditorStore );
+
+	const ViewSwitcherComponent = () => (
```
Comment (senadir) [inline src/components/use-view-switcher/index.js]: We updated ViewSwticher since you copied this.
```suggestion
	const ViewSwitcherComponent => (
```

File: src/edit.js [hunk 14]
```diff
@@ -2,43 +2,163 @@
...
+					'free-shipping-progress-bar'
+				) }
+				showErrorMessage={ true }
+				errorMessagePrefix={ __(
+					'Error message:',
+					'free-shipping-progress-bar'
+				) }
+			>
+				<BlockControls __experimentalShareWithChildBlocks>
+					<ViewSwitcherComponent />
```
Comment (senadir) [inline src/edit.js]:
```suggestion
					{ ViewSwitcherComponent }
```

**Review by senadir — APPROVED** [review]
This is looking good! Thank you for working on this.

We have a lot of reused coded from blocks here, we should look (later) into exporting it.

File: package.json (hunk 12 above)
Comment (nielslange) [reply package.json]: Moved! 😀

**Comment by nielslange** [comment]
> This is looking good! Thank you for working on this.

Thanks for your review and approving these changes.

> We have a lot of reused coded from blocks here, we should look (later) into exporting it.

Absolutely. The two biggest parts I reused are the `<BlockErrorBoundary>` and the `<useViewSwitcher>` components.
